File: Engine/Android/core/src/test/kotlin/despia/engine/BackgroundConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The background-task conformance runner - executes
 * OpenSource/Conformance/background/{declaration,constraints,budget}.json through THIS
 * runtime's BackgroundPlan (parity/F08-background.md). The TS twin (@despia/kernel
 * background.ts, background-conformance.test.ts) and the Swift reference (Engine/iOS
 * BackgroundPlan) run the SAME files, so a declared task cannot mean one thing on one
 * renderer and something else on another: the same clamp, the same constraint translation,
 * the same budget, and the same honest run record when the OS kills the process mid-flight.
 *
 * Missing corpus = loud failure - a silently-skipped conformance suite is how drift starts.
 */
class BackgroundConformanceTest {

    private fun corpusFile(name: String): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/background/$name.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/background/$name.json not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun root(name: String): Map<String, Any?> {
        val root = json(corpusFile(name).readText()).foundationValue as? Map<String, Any?>
            ?: error("$name.json: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "$name.json version")
        return root
    }

    private fun strings(value: Any?): List<String> =
        (value as? List<Any?>)?.map { it as String } ?: emptyList()

    @Suppress("UNCHECKED_CAST")
    private fun stringMap(value: Any?): Map<String, String> =
        (value as? Map<String, Any?>)?.entries?.associate { (k, v) -> k to (v as String) } ?: emptyMap()

    @Test
    fun vocabularyAgreesWithCorpus() {
        val doc = root("declaration")
        assertEquals(strings(doc["kinds"]), BackgroundPlan.KINDS, "kinds")
        assertEquals(strings(doc["requirements"]), BackgroundPlan.REQUIREMENTS, "requirements")
        assertEquals(strings(doc["buckets"]), BackgroundPlan.BUCKETS, "buckets")
        assertEquals(
            (doc["periodicFloorSeconds"] as Number).toInt(), BackgroundPlan.PERIODIC_FLOOR_SECONDS,
            "periodic floor",
        )
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun declarationFoldAgreesWithCorpus() {
        val cases = root("declaration")["declare"] as? List<Map<String, Any?>>
            ?: error("declaration.json: no declare[]")
        assertTrue(cases.isNotEmpty(), "declare corpus must not be empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "<unnamed>"
            val row = (case["row"] as? Map<String, Any?>)?.entries
                ?.associate { (k, v) -> k to (v as? String) } ?: emptyMap()
            val result = BackgroundPlan.resolveTask(
                case["id"] as? String, row, strings(case["declaredActions"]),
            )
            val expect = case["expect"] as? Map<String, Any?> ?: error("$name: no expect")

            val expectedError = expect["error"] as? String
            if (expectedError != null) {
                assertTrue(result.isFailure, "$name: expected refusal $expectedError, got ${result.getOrNull()}")
                val failure = result.exceptionOrNull() as? BackgroundPlan.RefusalError
                assertEquals(expectedError, BackgroundPlan.code(failure!!.refusal), "$name: refusal code")
                continue
            }

            assertTrue(result.isSuccess, "$name: expected a resolved task, got ${result.exceptionOrNull()?.message}")
            val task = result.getOrThrow()
            assertEquals(case["id"] as String, task.id, "$name: id")
            assertEquals(expect["action"] as String, task.action, "$name: action")
            assertEquals(expect["kind"] as String, task.kind, "$name: kind")
            assertEquals(
                (expect["minIntervalSeconds"] as Number).toInt(), task.minIntervalSeconds,
                "$name: minIntervalSeconds",
            )
            assertEquals(expect["clamped"] as Boolean, task.clamped, "$name: clamped")
            assertEquals(strings(expect["requires"]), task.requires, "$name: requires")
            assertEquals(expect["expedited"] as Boolean, task.expedited, "$name: expedited")
            assertEquals(expect["bucket"] as String, task.bucket, "$name: bucket")
            assertTrue(task.bucket in BackgroundPlan.BUCKETS, "$name: bucket is one of the fixed five")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun constraintTranslationAgreesWithCorpus() {
        val rows = root("constraints")["map"] as? List<Map<String, Any?>>
            ?: error("constraints.json: no map[]")
        assertEquals(BackgroundPlan.REQUIREMENTS.size, rows.size, "every requirement is mapped")
        for (row in rows) {
            val requirement = row["requirement"] as String
            val folded = BackgroundPlan.constraints(listOf(requirement))
            assertEquals(stringMap(row["ios"]), folded.ios, "$requirement: iOS constraints")
            assertEquals(stringMap(row["android"]), folded.android, "$requirement: Android constraints")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun constraintFoldAgreesWithCorpus() {
        val cases = root("constraints")["fold"] as? List<Map<String, Any?>>
            ?: error("constraints.json: no fold[]")
        assertTrue(cases.isNotEmpty(), "fold corpus must not be empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "<unnamed>"
            val folded = BackgroundPlan.constraints(strings(case["requires"]))
            assertEquals(stringMap(case["ios"]), folded.ios, "$name: iOS")
            assertEquals(stringMap(case["android"]), folded.android, "$name: Android")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun constraintSatisfactionAgreesWithCorpus() {
        val cases = root("constraints")["satisfy"] as? List<Map<String, Any?>>
            ?: error("constraints.json: no satisfy[]")
        assertTrue(cases.isNotEmpty(), "satisfy corpus must not be empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "<unnamed>"
            val stateRow = case["state"] as? Map<String, Any?> ?: emptyMap()
            val state = BackgroundPlan.DeviceState(
                network = stateRow["network"] as? String ?: "none",
                charging = stateRow["charging"] as? Boolean ?: false,
                batteryLow = stateRow["batteryLow"] as? Boolean ?: false,
                storageLow = stateRow["storageLow"] as? Boolean ?: false,
                idle = stateRow["idle"] as? Boolean ?: false,
            )
            val expect = case["expect"] as? Map<String, Any?> ?: error("$name: no expect")
            val unmet = BackgroundPlan.unmet(strings(case["requires"]), state)
            assertEquals(strings(expect["unmet"]), unmet, "$name: unmet")
            assertEquals(expect["satisfied"] as Boolean, unmet.isEmpty(), "$name: satisfied")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun budgetCountdownAgreesWithCorpus() {
        val doc = root("budget")
        val budgets = (doc["budgets"] as? Map<String, Any?>)?.entries
            ?.associate { (k, v) -> k to (v as Number).toLong() } ?: emptyMap()
        assertEquals(budgets, BackgroundPlan.BUDGET_MS, "budgets")

        val cases = doc["countdown"] as? List<Map<String, Any?>> ?: error("budget.json: no countdown[]")
        assertTrue(cases.isNotEmpty(), "countdown corpus must not be empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = case["expect"] as? Map<String, Any?> ?: error("$name: no expect")
            val budget = BackgroundPlan.budget(
                case["platform"] as String, (case["elapsedMs"] as Number).toLong(),
            )
            assertEquals((expect["remainingMs"] as Number).toLong(), budget.remainingMs, "$name: remainingMs")
            assertEquals(expect["expired"] as Boolean, budget.expired, "$name: expired")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun runRecordAgreesWithCorpus() {
        val cases = root("budget")["record"] as? List<Map<String, Any?>> ?: error("budget.json: no record[]")
        assertTrue(cases.isNotEmpty(), "record corpus must not be empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = case["expect"] as? Map<String, Any?> ?: error("$name: no expect")
            val record = BackgroundPlan.runRecord(strings(case["events"]))
            assertEquals(expect["result"] as String, record.result, "$name: result")
            assertEquals(expect["failure"] as Boolean, record.failure, "$name: failure")
            assertEquals(expect["running"] as Boolean, record.running, "$name: running")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun debugGateAgreesWithCorpus() {
        val cases = root("budget")["debugGate"] as? List<Map<String, Any?>>
            ?: error("budget.json: no debugGate[]")
        assertTrue(cases.isNotEmpty(), "debugGate corpus must not be empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "<unnamed>"
            val expect = case["expect"] as? Map<String, Any?> ?: error("$name: no expect")
            val allowed = BackgroundPlan.runAllowed(case["channel"] as? String)
            assertEquals(expect["allowed"] as Boolean, allowed, "$name: allowed")
            if (expect["allowed"] == false) {
                assertEquals("debug_only", expect["error"] as String, "$name: the refusal is typed")
            }
        }
    }
}

File: Web/packages/kernel/test/background-conformance.test.ts
//
//  background-conformance.test.ts - the SHARED background corpus through the TS kernel
//  (OpenSource/Conformance/background/{declaration,constraints,budget}.json). The Kotlin twin
//  (:core BackgroundConformanceTest) and the Swift twin (Engine/iOS BackgroundPlan) run the
//  SAME files, so a declared task cannot mean one thing on one renderer and something else on
//  another: the same clamp, the same constraint translation, the same budget, and the same
//  honest run record when the OS kills the process mid-flight.
//
//  Missing corpus = loud failure - a silently-skipped conformance suite is how drift starts.
//
//  Imported from ../src/background.ts directly rather than through the package barrel: the
//  barrel is a shared file this workstream does not own (parity/AGENT-CONTRACT.md), and the
//  export line rides in the handoff.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  BACKGROUND_KINDS, BACKGROUND_REQUIREMENTS, BACKGROUND_PERIODIC_FLOOR_SECONDS,
  BACKGROUND_BUCKETS, BACKGROUND_BUDGET_MS,
  resolveBackgroundTask, backgroundConstraints, backgroundUnmet, backgroundBudget,
  backgroundRunRecord, backgroundRunAllowed,
  type BackgroundRow, type BackgroundRunEvent, type BackgroundDeviceState,
} from "../src/background.ts";

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/background");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/background not found");
    dir = parent;
  }
}

function corpus(name: string): { [k: string]: unknown } {
  const doc = JSON.parse(readFileSync(join(corpusDir(), `${name}.json`), "utf-8")) as { [k: string]: unknown };
  assert.equal(doc["version"], 1, `${name}.json: version`);
  return doc;
}

test("background: the declared vocabulary agrees with the corpus", () => {
  const doc = corpus("declaration");
  assert.deepEqual([...BACKGROUND_KINDS], doc["kinds"]);
  assert.deepEqual([...BACKGROUND_REQUIREMENTS], doc["requirements"]);
  assert.deepEqual([...BACKGROUND_BUCKETS], doc["buckets"]);
  assert.equal(BACKGROUND_PERIODIC_FLOOR_SECONDS, doc["periodicFloorSeconds"]);
});

test("background: the declaration fold agrees with the corpus", () => {
  type Case = {
    name: string; id: string; declaredActions: string[]; row: BackgroundRow;
    expect: { [k: string]: unknown };
  };
  const cases = corpus("declaration")["declare"] as Case[];
  assert.ok(cases.length > 0, "declare corpus must not be empty");
  for (const testCase of cases) {
    const result = resolveBackgroundTask(testCase.id, testCase.row, testCase.declaredActions);
    const expected = testCase.expect;

    if ("error" in expected) {
      assert.equal(result.ok, false, `${testCase.name}: expected refusal ${String(expected["error"])}`);
      assert.equal(result.ok === false && result.error, expected["error"], `${testCase.name}: refusal code`);
      continue;
    }

    assert.equal(result.ok, true, `${testCase.name}: expected a resolved task`);
    if (result.ok !== true) continue;
    const task = result.value;
    assert.equal(task.id, testCase.id, `${testCase.name}: id`);
    assert.equal(task.action, expected["action"], `${testCase.name}: action`);
    assert.equal(task.kind, expected["kind"], `${testCase.name}: kind`);
    assert.equal(task.minIntervalSeconds, expected["minIntervalSeconds"], `${testCase.name}: minIntervalSeconds`);
    assert.equal(task.clamped, expected["clamped"], `${testCase.name}: clamped`);
    assert.deepEqual([...task.requires], expected["requires"], `${testCase.name}: requires`);
    assert.equal(task.expedited, expected["expedited"], `${testCase.name}: expedited`);
    assert.equal(task.bucket, expected["bucket"], `${testCase.name}: bucket`);
    assert.ok(BACKGROUND_BUCKETS.includes(task.bucket), `${testCase.name}: bucket is one of the fixed five`);
  }
});

test("background: each requirement translates to the platform constraint the corpus pins", () => {
  type Row = { requirement: string; ios: { [k: string]: string }; android: { [k: string]: string } };
  const rows = corpus("constraints")["map"] as Row[];
  assert.equal(rows.length, BACKGROUND_REQUIREMENTS.length, "every requirement is mapped");
  for (const row of rows) {
    const folded = backgroundConstraints([row.requirement]);
    assert.deepEqual(folded.ios, row.ios, `${row.requirement}: iOS constraints`);
    assert.deepEqual(folded.android, row.android, `${row.requirement}: Android constraints`);
  }
});

test("background: constraint sets fold as the corpus pins", () => {
  type Case = { name: string; requires: string[]; ios: { [k: string]: string }; android: { [k: string]: string } };
  const cases = corpus("constraints")["fold"] as Case[];
  assert.ok(cases.length > 0, "fold corpus must not be empty");
  for (const testCase of cases) {
    const folded = backgroundConstraints(testCase.requires);
    assert.deepEqual(folded.ios, testCase.ios, `${testCase.name}: iOS`);
    assert.deepEqual(folded.android, testCase.android, `${testCase.name}: Android`);
  }
});

test("background: a task does not run while a constraint is unmet", () => {
  type Case = {
    name: string; requires: string[]; state: BackgroundDeviceState;
    expect: { satisfied: boolean; unmet: string[] };
  };
  const cases = corpus("constraints")["satisfy"] as Case[];
  assert.ok(cases.length > 0, "satisfy corpus must not be empty");
  for (const testCase of cases) {
    const unmet = backgroundUnmet(testCase.requires, testCase.state);
    assert.deepEqual(unmet, testCase.expect.unmet, `${testCase.name}: unmet`);
    assert.equal(unmet.length === 0, testCase.expect.satisfied, `${testCase.name}: satisfied`);
  }
});

test("background: the budget countdown agrees with the corpus", () => {
  const doc = corpus("budget");
  assert.deepEqual({ ...BACKGROUND_BUDGET_MS }, doc["budgets"]);

  type Case = { name: string; platform: string; elapsedMs: number; expect: { remainingMs: number; expired: boolean } };
  const cases = doc["countdown"] as Case[];
  assert.ok(cases.length > 0, "countdown corpus must not be empty");
  for (const testCase of cases) {
    const budget = backgroundBudget(testCase.platform, testCase.elapsedMs);
    assert.equal(budget.remainingMs, testCase.expect.remainingMs, `${testCase.name}: remainingMs`);
    assert.equal(budget.expired, testCase.expect.expired, `${testCase.name}: expired`);
  }
});

test("background: the run record agrees with the corpus, kills included", () => {
  type Case = {
    name: string; events: BackgroundRunEvent[];
    expect: { result: string; failure: boolean; running: boolean };
  };
  const cases = corpus("budget")["record"] as Case[];
  assert.ok(cases.length > 0, "record corpus must not be empty");
  for (const testCase of cases) {
    const record = backgroundRunRecord(testCase.events);
    assert.equal(record.result, testCase.expect.result, `${testCase.name}: result`);
    assert.equal(record.failure, testCase.expect.failure, `${testCase.name}: failure`);
    assert.equal(record.running, testCase.expect.running, `${testCase.name}: running`);
  }
});

test("background: run is refused in release, and fails closed", () => {
  type Case = { name: string; channel: string; expect: { allowed: boolean; error?: string } };
  const cases = corpus("budget")["debugGate"] as Case[];
  assert.ok(cases.length > 0, "debugGate corpus must not be empty");
  for (const testCase of cases) {
    const allowed = backgroundRunAllowed(testCase.channel);
    assert.equal(allowed, testCase.expect.allowed, `${testCase.name}: allowed`);
    if (testCase.expect.allowed === false) {
      assert.equal(testCase.expect.error, "debug_only", `${testCase.name}: the refusal is typed`);
    }
  }
});
